### PR TITLE
[DataGrid] Add separate density factor for column headers

### LINKED ISSUE #5274: [data grid] Improve header behavior with density property
### Summary 💡

The header should respond better to the density property. It should feel compact, but not in the detriment of the elements we use on the header.

Currently, with the new Aggregation implementation feels a bit to tight.

image

### Examples 🌈

- Gmail density:
Tabs (comparable to our headers) get more compact, but less than the data rows

### Motivation 🔦

- Behavior to better support the aggregation function and other elements we can use on the header.

### Order ID 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by KenanYusuf**
Adds a separate density factor for column headers. This allows us to control the size of columns headers independently from rows.

The `compact` density for column headers with this change is `0.8` rather than the `0.7` it was before. This means that column headers appear slightly taller now with the compact density set, buying us more space for other UI elements in the column headers like the aggregation label.

Fixes #5274

**Before**

Screenshot 2024-07-29 at 16 21 49

**After**

Screenshot 2024-07-29 at 16 21 33

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/x-data-grid/src/hooks/features/density/densitySelector.ts [hunk 1]
```diff
@@ -4,16 +4,29 @@ import { GridDensity } from '../../../models/gridDensity';
 
 export const COMPACT_DENSITY_FACTOR = 0.7;
 export const COMFORTABLE_DENSITY_FACTOR = 1.3;
+export const COMPACT_DENSITY_FACTOR_HEADER = 0.8;
+export const COMFORTABLE_DENSITY_FACTOR_HEADER = 1.3;
```
Comment (KenanYusuf) [inline packages/x-data-grid/src/hooks/features/density/densitySelector.ts:8]: We may want to lower this at some point, as `comfortable` column headers are quite tall.

File: packages/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts [hunk 2]
```diff
@@ -91,9 +91,10 @@ export function useGridDimensions(
   const pinnedColumns = useGridSelector(apiRef, gridVisiblePinnedColumnDefinitionsSelector);
   const densityFactor = useGridSelector(apiRef, gridDensityFactorSelector);
   const rowHeight = Math.floor(props.rowHeight * densityFactor);
-  const headerHeight = Math.floor(props.columnHeaderHeight * densityFactor);
+  const headerDensityFactor = useGridSelector(apiRef, gridHeaderDensityFactorSelector);
+  const headerHeight = Math.floor(props.columnHeaderHeight * headerDensityFactor);
   const headerFilterHeight = Math.floor(
-    (props.headerFilterHeight ?? props.columnHeaderHeight) * densityFactor,
+    (props.headerFilterHeight ?? props.columnHeaderHeight) * headerDensityFactor,
```
Comment (KenanYusuf) [inline packages/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts:97]: We have a separate issue to deal with the fact that compact header filters are overflowing; this change neither fixes or worsens this issue https://github.com/mui/mui-x/issues/13048

**Review by romgrk — COMMENTED** [review]
I don't think I like having a separate density for headers. Is there any other way to solve the issue? Could we adjust the font-size for the label instead?

**Comment by KenanYusuf** [comment]
> I don't think I like having a separate density for headers. Is there any other way to solve the issue? Could we adjust the font-size for the label instead?

@romgrk I don't think reducing the font-size is the right solution here as it would lower the importance of the labels. 

We need to ensure that the header rows have enough space to accommodate all of the elements that can be rendered in it - another option is just setting a minimum height for the header row...

**Comment by romgrk** [comment]
How about vertically centered header's content instead of aligning on the title's baseline?

image

> another option is just setting a minimum height for the header row...

I also don't like it, it would still be a separate header height.


@mui/xgrid Any other opinions on this issue? I don't like separate height but I'll go with the consensus if the rest of the team is fine with it.

**Comment by KenanYusuf** [comment]
> How about vertically centered header's content instead of aligning on the title's baseline?

We have discussed this option in the [Figma explorations](https://www.figma.com/design/rPwfaFijKseEDWpJYg46rA?node-id=45-82#872189208) - I don't love the misalignment but it would look better if we had persistent separators between the column headers like your example shows.

**Comment by romgrk** [comment]
I'd be happy with persistent separators, I find our current ones a bit jarring.

**Pull request closed without merging**